feat(app-config): pwa validation

Francois-Esquire opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Uniform validation for `appConfig`, starting with PWA config. The old PWA config validation has been removed in favor of the `joi` schema.

## Description
<!--- Describe your changes in detail -->
Using validation library `joi`, a PWA schema was created for that portion of `appConfig` and wrapped with a validation function `validatePWAConfig`.

Files added:
- validation schemas under `src/server/utils/validation` + tests.

Files removed:
- `src/server/middleware/pwa/validation.js` + test

Files modified (w/ tests if applicable):
- `src/server/utils/onModuleLoad.js`
- `src/server/middleware/pwa/config.js`

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Utilize a single point of validation for user specific settings that plug into and configure the app.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Testing suite

## Types of Changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (adding or updating documentation)
- [ ] Dependency update

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] These changes should be applied to a maintenance branch.
- [ ] This change requires cross browser checks.
- [ ] Performance tests should be ran against the server prior to merging.
- [ ] This change impacts caching for client browsers.
- [ ] This change impacts HTTP headers.
- [ ] This change adds additional environment variable requirements for One App users.
- [x] I have added the Apache 2.0 license header to any new files created.

## What is the Impact to Developers Using One App?
<!--- Please describe how your changes impacts developers using One App. -->
Richer error messages that identify problem areas in `appConfig`

JamesSingleton commented on src/server/utils/validation/index.js:
```diff
@@ -0,0 +1,75 @@
+  pwa: pwaSchema,
+  createSsrFetch: Joi.function(),
+  eventLoopDelayThreshold: Joi.number().allow(Infinity),
+  appCompatibility: semverExtension.semver().valid(),
```
Doesn't this come in as 

```json
"app": {
  "compatibility : "x.x.x"
}
```

I forget where the conversion to `appCompatibility` comes into play.

Francois-Esquire replied:
Yes! `@americanexpress/one-app-bundler` actually sets this on the modules meta data. When we come to validate it, we can access it from the module and make sure it's compatible with the running One App instance.

mtomcal commented on __tests__/server/utils/validation/index.spec.js:
```diff
@@ -0,0 +1,57 @@
+      new Error([
+        '"recoveryMode" must be a boolean',
+        '"escapeHatch" must be a boolean',
+        '"scope" with value "\\" fails to match the required pattern: /^\\//',
```
Can we get a more humanized message for `scope`? https://joi.dev/api/?v=17.2.1#anyruleoptions

mtomcal commented on src/server/utils/validation/webManifest.js:
```diff
@@ -0,0 +1,83 @@
+  scope: Joi.string().pattern(/^\//),
+  start_url: Joi.alternatives(
+    Joi.string().uri(),
+    Joi.string().pattern(/^\//)
```
Some more humanized messages for `pattern` would be good.

Francois-Esquire commented:
@mtomcal I've updated the error messages for `scope` and `start_url` ( a040edc ),  it's a friendlier message.  I've also added more samples to the test to view the output; let me know if we can write/communicate a specific field better.
